cabal-install/main/Main.hs

```diff
@@ -313,18 +314,27 @@ mainWorker args = topHandler $
       , hiddenCmd  actAsSetupCommand actAsSetupAction
       , hiddenCmd  manpageCommand (manpageAction commandSpecs)
 
-      , regularCmd  CmdConfigure.configureCommand CmdConfigure.configureAction
-      , regularCmd  CmdBuild.buildCommand         CmdBuild.buildAction
-      , regularCmd  CmdRepl.replCommand           CmdRepl.replAction
-      , regularCmd  CmdFreeze.freezeCommand       CmdFreeze.freezeAction
-      , regularCmd  CmdHaddock.haddockCommand     CmdHaddock.haddockAction
-      , regularCmd  CmdRun.runCommand             CmdRun.runAction
-      , regularCmd  CmdTest.testCommand           CmdTest.testAction
-      , regularCmd  CmdBench.benchCommand         CmdBench.benchAction
+      , regularCmd  CmdConfigure.configureCommand (defaultsCmdWrapper CmdConfigure.configureAction)
+      , regularCmd  CmdBuild.buildCommand         (defaultsCmdWrapper CmdBuild.buildAction)
+      , regularCmd  CmdRepl.replCommand           (defaultsCmdWrapper CmdRepl.replAction)
+      , regularCmd  CmdFreeze.freezeCommand       (defaultsCmdWrapper CmdFreeze.freezeAction)
+      , regularCmd  CmdHaddock.haddockCommand     (defaultsCmdWrapper CmdHaddock.haddockAction)
+      , regularCmd  CmdRun.runCommand             (defaultsCmdWrapper CmdRun.runAction)
+      , regularCmd  CmdTest.testCommand           (defaultsCmdWrapper CmdTest.testAction)
+      , regularCmd  CmdBench.benchCommand         (defaultsCmdWrapper CmdBench.benchAction)
       ]
 
 type Action = GlobalFlags -> IO ()
 
+defaultsCmdWrapper :: ((ConfigFlags, ConfigExFlags, InstallFlags, HaddockFlags) -> [String] -> GlobalFlags -> IO ())
```

Comment:
@23Skidoo we could use ViewPatterns, then it would look like

```hs
buildAction (cliFlags@(configFlags, configExFlags, installFlags, haddockFlags) <- applyFlagDefaults)
```